serverless/resource_serverless.go

```diff
@@ -0,0 +1,158 @@
+package serverless
+
+import (
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceServerless() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceServerlessCreate,
+		Read:   resourceServerlessRead,
+		Update: resourceServerlessUpdate,
+		Delete: resourceServerlessDelete,
+
+		Schema: map[string]*schema.Schema{
+			"config_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"serverless_bin_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			// The directory where the Serverless package lives. In the CLI, this defaults to
+			// .serverless, but we default to .terraform-serverless to avoid an issue where
+			// the CLI deletes the .serverless directory after deploy, even with --package.
+			// Note that the provider requires out-of-band packaging. Users should package
+			// their code with `sls package --package .terraform-serverless`.
+			//
+			// NOTE: the path you provide must be RELATIVE to your `config_dir` since the
+			// --package flag in the CLI does not support absolute paths.
+			"package_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  ".terraform-serverless",
+			},
+			"stage": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"args": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"package_hash": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+		},
+
+		// Only trigger a deploy if either the Serverless config or Serverless zip archive has changed.
+		// `sls package` isn't deterministic according to experiments, so in practive this means that
```

Comment:
Minor: `s/practive/practice/`

Reply:
Minor note for us -- I _think_ serverless attempts to be deterministic on builds. Part of that behavior involves wiping mtime and setting all resources to epoch when zipping.

Jetpack doesn't do this (probably what @tptee is working with). We have a [jetpack ticket](https://github.com/FormidableLabs/serverless-jetpack/issues/7) to consider adding mtime nuking (and possibly other deterministic build things) and if we think that would help our efforts here (or generally), I'm happy to start looking into that ticket (theoretically shouldn't be that difficult).

Reply:
Captured in https://github.com/FormidableLabs/terraform-provider-serverless/issues/10